File: 백준/Gold/3055. 탈출/탈출.cc
#include <bits/stdc++.h>
using namespace std;

int r, c;
char mat[51][51];
int dist[51][51];
int waterDist[51][51];
int visited[51][51];
pair<int, int> dest;

int dx[] = {1, -1, 0, 0}, dy[] = {0, 0, 1, -1};

void flood()
{
    fill(&waterDist[0][0], &waterDist[0][0] + 51 * 51, -1);
    queue<pair<int, int>> q;
    for (int i = 0; i < r; i++)
    {
        for (int j = 0; j < c; j++)
        {
            if (mat[i][j] == '*')
            {
                q.push({i, j});      // 물인 칸 큐에 넣기
                waterDist[i][j] = 0; // 물 시작 거리 0
                visited[i][j] = 1;   // 물 방문 했나 안했나 확인
            }
        }
    }

    while (q.size())
    {
        int x, y;
        tie(x, y) = q.front();
        q.pop();

        for (int i = 0; i < 4; i++)
        {
            int nx = x + dx[i];
            int ny = y + dy[i];

            if (nx < 0 || nx >= r || ny < 0 || ny >= c)
                continue; // 좌표 범위 초과
            if (mat[nx][ny] != '.')
                continue; // 물이 넘칠 수 있는 칸은 .인 빈칸 뿐
            if (visited[nx][ny])
                continue; // 이미 방문한 곳은 패스

            waterDist[nx][ny] = waterDist[x][y] + 1;
            q.push({nx, ny});
            visited[nx][ny] = 1;
        }
    }
}

void move()
{
    fill(&dist[0][0], &dist[0][0] + 51 * 51, -1);
    queue<pair<int, int>> q;
    for (int i = 0; i < r; i++)
    {
        for (int j = 0; j < c; j++)
        {
            visited[i][j] = 0; // visited 위에서 물 이동때 사용한 것 재사용하기 위해 모두 0으로 초기화하는 줄
            if (mat[i][j] == 'S')
            { // 시작점 찾기
                q.push({i, j});
                dist[i][j] = 0;
                visited[i][j] = 1;
            }
            else if (mat[i][j] == 'D')
            {
                dest = {i, j};
            }
        }
    }

    while (q.size())
    {
        int x, y;
        tie(x, y) = q.front();
        q.pop();

        for (int i = 0; i < 4; i++)
        {
            int nx = x + dx[i];
            int ny = y + dy[i];

            if (nx < 0 || nx >= r || ny < 0 || ny >= c)
                continue; // 좌표 범위 초과
            if (mat[nx][ny] != '.' and mat[nx][ny] != 'D')
                continue; // 물이 넘칠 수 있는 칸은 .인 빈칸 뿐, 도착점은 가야함
            if (waterDist[nx][ny] != -1 && waterDist[nx][ny] <= dist[x][y] + 1)
                continue; // 물에 잠긴 곳 혹은 다음 턴에 물에 잠길 곳은 패스
            if (visited[nx][ny])
                continue; // 방문한 곳 다시 안가기

            dist[nx][ny] = dist[x][y] + 1;
            q.push({nx, ny});
            visited[nx][ny] = 1;
        }
    }
}

void getInput()
{
    cin >> r >> c;
    for (int i = 0; i < r; i++)
    {
        for (int j = 0; j < c; j++)
        {
            cin >> mat[i][j];
        }
    }
}

int main()
{
    ios::sync_with_stdio(0);
    cin.tie(NULL);

    getInput();

    flood();
    move();

    if (dist[dest.first][dest.second] != -1)
    {
        cout << dist[dest.first][dest.second] << "\n";
    }
    else
    {
        cout << "KAKTUS"
             << "\n";
    }
    return 0;
}

File: 백준/Gold/3055. 탈출/탈출.py
import sys 
input = sys.stdin.readline

from collections import deque

n, m = map(int, input().split());
mat = [input().strip() for _ in range(n)]
visit = [[0] * m for _ in range(n)]
dist = [[-1] * m for _ in range(n)]
water_dist = [[-1] * m for _ in range(n)]


def flood():
    q = deque()
    for i in range(n):
        for j in range(m):
            if mat[i][j] == '*':
                q.append((i, j))
                visit[i][j] = 1
                water_dist[i][j] = 0
    
    while q:
        x, y = q.popleft()
        for dx, dy in zip([1, -1, 0, 0], [0, 0, 1, -1]):
            nx, ny = x + dx, y + dy
            if nx < 0 or nx >= n or ny < 0 or ny >= m:  # 좌표 범위 초과
                continue
            if mat[nx][ny] != '.':  # 빈칸이 아니면 물이 넘칠 수 없음
                continue
            if visit[nx][ny]:  # 이미 방문한 곳은 안감
                continue

            q.append((nx, ny))
            visit[nx][ny] = 1
            water_dist[nx][ny] = water_dist[x][y] + 1  # 현재 칸에 물이 몇 번째 턴에 차는지 기록 -> 이후 고슴도치 이동 시 사용


def move():
    q = deque()
    for i in range(n):
        for j in range(m):
            visit[i][j] = 0
            if mat[i][j] == 'S':
                q.append((i, j))
                visit[i][j] = 1
                dist[i][j] = 0

    while q:
        x, y = q.popleft()
        for dx, dy in zip([1, -1, 0, 0], [0, 0, 1, -1]):
            nx, ny = x + dx, y + dy
            if nx < 0 or nx >= n or ny < 0 or ny >= m:  # 좌표 범위 초과
                continue
            if mat[nx][ny] != "." and mat[nx][ny] != 'D':  # 빈칸이 아니거나 비버굴이 아니면 방문 안해
                continue
            if water_dist[nx][ny] != -1 and water_dist[nx][ny] <= dist[x][y] + 1:  # 물이 이미 넘친곳이거나 다음 턴에 물에 잠길 곳
                continue
            if visit[nx][ny]:  # 이미 방문한 좌표
                continue

            q.append((nx, ny))
            visit[nx][ny] = 1
            dist[nx][ny] = dist[x][y] + 1


flood()  # 물 먼저 이동
move()   # 고슴도치 이동

for i in range(n):
    for j in range(m):
        if mat[i][j] == 'D':
            if dist[i][j] == -1:
                print("KAKTUS")
            else:
                print(dist[i][j])